lib/str.d.ts

```diff
@@ -0,0 +1,643 @@
+     * @param {string} input
+     * @returns {string}
+     */
+    removeSpaces(input: string): string;
+    /**
+     * Returns the proper phrase depending on the count that is passed.
+     * Example:
+     * console.log(Str.pluralize('puppy', 'puppies', 1)); // puppy
+     * console.log(Str.pluralize('puppy', 'puppies', 3)); // puppies
+     *
+     * @param {String} singular form of the phrase
+     * @param {String} plural form of the phrase
+     * @param {Number} n the count which determines the plurality
+     *
+     * @return {String}
+     */
+    pluralize(singular: string, plural: string, n: number): string;
+    /**
+     * Returns whether or not a string is an encrypted number or not.
+     *
+     * @param {String} number that we want to see if its encrypted or not
+     *
+     * @return {Boolean} Whether or not this string is an encrpypted number
+     */
+    isEncryptedCardNumber(number: string): boolean;
+    /**
+     * Converts a value to boolean, case-insensitive.
+     * @param {mixed} value
+     * @return {Boolean}
+     */
+    toBool(value: unknown): boolean;
+    /**
+     * Checks if a string could be the masked version of another one.
+     *
+     * @param {String} first string to compare
+     * @param {String} second string to compare
+     * @param {String} [mask] defaults to X
+     * @return {Boolean} true when first could be the masked version of second
+     */
+    maskedEquals(first: string, second: string, mask: string): boolean;
+    /**
+     * Bold any word matching the regexp in the text.
+     * @param {string} text, htmlEncoded
+     * @param {RegExp} regexp
+     * @return {string}
+     */
+    boldify(text: string, regexp: RegExp): string;
+    /**
+     * Check for whether a phone number is valid.
+     * @param {String} phone
+     *
+     * @return {bool}
+     */
+    isValidPhone(phone: string): boolean;
+    /**
+     * We validate mentions by checking if it's first character is an allowed character.
+     *
+     * @param {String} mention
+     * @returns {bool}
+     */
+    isValidMention(mention: string): boolean;
+    /**
+     * Returns text without our SMS domain
+     *
+     * @param {String} text
+     * @return {String}
+     */
+    removeSMSDomain(text: string): string;
+    /**
+     * Returns true if the text is a valid phone number with our SMS domain removed
+     *
+     * @param {String} text
+     * @return {bool}
+     */
+    isSMSLogin(text: string): boolean;
+    /**
+     * This method will return all matches of a single regex like preg_match_all() in PHP. This is not a common part of
+     * JS yet, so this is a good way of doing it according to
+     * https://github.com/airbnb/javascript/issues/1439#issuecomment-306297399 and doesn't get us in trouble with
+     * linting rules.
+     *
+     * @param {String} str
+     * @param {RegExp} regex
+     *
+     * @returns {Array}
+     */
+    matchAll(str: string, regex: RegExp): unknown[];
```

Comment:
Looking at the library code, you'll realize that each match will always have `input` and `index`. 
Also, depending of the number of capturing groups te passed regex has, it has `[key: number]: string` mapping. So we can specify the return type more specifically than `unknonw[]`

`...regex: RegExp): Array<{ [key: number]: string; input: string; index: number }>;`

Reply:
I just realized we can also do something like this

```
...regex: RegExp): Array<RegExpMatchArray & { input: string; index: number }>;
```